htlcswitch/payment_result.go

```diff
@@ -297,3 +425,84 @@ func (store *networkResultStore) CleanStore(keep map[uint64]struct{}) error {
 		return nil
 	}, func() {})
 }
+
+// FetchPendingAttempts returns a list of all attempt IDs that are currently in
+// the pending state.
+func (store *networkResultStore) FetchPendingAttempts() ([]uint64, error) {
+	var pending []uint64
+	err := kvdb.View(store.backend, func(tx kvdb.RTx) error {
+		bucket := tx.ReadBucket(networkResultStoreBucketKey)
+		if bucket == nil {
+			return nil
+		}
+
+		return bucket.ForEach(func(k, v []byte) error {
+			// If the key is not 8 bytes, it's not a valid attempt
+			// ID.
+			if len(k) != 8 {
+				return nil
+			}
+
+			// Deserialize the result to check its type.
+			r := bytes.NewReader(v)
+			result, err := deserializeNetworkResult(r)
+			if err != nil {
+				// If we can't deserialize, we'll log it and
+				// continue.
+				log.Warnf("Unable to deserialize result for "+
+					"key %x: %v", k, err)
+				return nil
+			}
+
+			// If the result is a pending result, add the attempt
+			// ID to our list.
+			if result.msg.MsgType() == pendingHtlcMsgType {
+				attemptID := binary.BigEndian.Uint64(k)
+				pending = append(pending, attemptID)
+			}
+
+			return nil
+		})
+	}, func() {
+		pending = nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return pending, nil
+}
+
+// FailAttempt transitions an initialized attempt from PENDING to FAILED,
+// recording the provided reason. This is the synchronous rollback mechanism for
+// attempts that fail before being committed to the forwarding engine. It
+// returns an error if the underlying storage fails.
+func (store *networkResultStore) FailAttempt(attemptID uint64,
+	linkErr *LinkError) error {
+
+	// The attempt to send the htlc failed before it was ever dispatched.
+	// We will write a failure result to the store to unblock any
+	// potential callers to GetAttemptResult.
+
+	// First, we need to serialize the wire message from our link error
+	// into a byte slice. This is what the downstream parsers expect.
+	var reasonBytes bytes.Buffer
+	wireMsg := linkErr.WireMessage()
+	if err := lnwire.EncodeFailure(&reasonBytes, wireMsg, 0); err != nil {
+		return fmt.Errorf("failed to encode failure for attempt %d: %w",
+			attemptID, err)
+	}
+
+	// We'll create a synthetic UpdateFailHTLC to represent this internal
+	// failure, following the pattern used by the contract resolver.
+	failMsg := &lnwire.UpdateFailHTLC{
+		Reason: lnwire.OpaqueReason(reasonBytes.Bytes()),
+	}
+
+	failureResult := &networkResult{
+		msg:         failMsg,
+		unencrypted: true, // This is a local failure
+	}
+
+	return store.StoreResult(attemptID, failureResult)
```

Comment:
as mentioned before I am hesitant to overwrite stuff hmm

Reply:
Per offline discussion, I added in a check to `FailAttempt` which ensures that it can only be used to update from an initialized attempt state to a failed state. This would provide some protection against its accidental use which could overwrite the real, final `settle/fail` result which arrives back from the network. Take a look and let me know if you think that's sufficient.